### ISSUE TITLE
Linux per-vlan STP state support

### ISSUE THREAD

**Issue by troglobit** [mstpd/mstpd#97]
Hi, has anyone seen the new per-vlan STP state support in Linux?

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a580c76d534c7360ba68042b19cb255e8420e987

We¹ are using the Linux bridge's native VLAN support, which is similar to how switch fabrics operate, with switchdev against Marvell switch cores. The new support for per-vlan stp/rstp, and I guess MSTP, is very interesting and something we could be able to help out with, unless someone else (Cumulus?) already have something lined up already? (Can't promise any timelines, but it's being actively talked about.)
__
¹ https://github.com/westermo

**Comment by commodo** [comment]
Looks interesting.
Thanks for pointing it out.
I [personally] don't follow the netdev list; so it takes a while until I find out these things.

I was wondering when [or if] this [per-VLAN STP] would become available in the kernel.
@vivien   maybe has some input? he was more involved with DSA at some point;

I will try to keep an eye for this.
Cumulus is already a user of mstpd ( see  http://oss.cumulusnetworks.com/CumulusLinux-2.5.12/patches/mstpd/ ), but they are using an older version with their own patches.

Depending on how the kernel STP evolves, this project will become more irrelevant.
Which is fine from my side.
Supporting baremetal OSes hasn't been too centric, and Linux was the main platform for mstpd.

Unfortunately, I don't have any time to allocate for with STP on Linux. Nor HW to test.
I guess Cumulus may be the main driver here, and they likely have the resources and interest to drive this.
And they also know mstpd [internally].

Hmmm, so I just found out that Nvidia acquired Cumulus recently; that sounds interesting.
https://techcrunch.com/2020/05/04/nvidia-acquires-cumulus-networks/
I guess that may mean more interest for these protocols in the kernel.
Let's see.

**Comment by troglobit** [comment]
From what I understand, talking to my colleague @wkz, STP in the kernel will remain for backwards compatibility reasons, but will not evolve into full RSTP or MSTP support.  Hence this revived `mstpd` project (yay! :) is the best candidate right now for anyone wanting anything more capable than plain STP.

From what I understand Cumulus moved away from their patches/oss to use plain .deb files.  So in the case of mstpd there's this, which seems to be current:

http://apt.cumulusnetworks.com/repo/pool/cumulus/m/mstpd/

I'll see what I can do to coordinate efforts on this from the Westermo side :)

**Comment by KanjiMonster** [comment]
What's the current state of this? Has anyone started working on this?

**Comment by commodo** [comment]
Hey,

Nope. There is no current work done on this [at least that I know of].

Thanks
Alex

**Comment by troglobit** [comment]
Nothing from our side yet either. Promise I'll update here when Westermo takes the plunge :)

**Comment by KanjiMonster** [comment]
Okay, good (?) to know. Thank you both for the updates.

**Comment by KanjiMonster** [comment]
So I did an attempt in https://github.com/mstpd/mstpd/pull/103 with a lot of open questions (for me).

**Comment by pavlix** [comment]
> Depending on how the kernel STP evolves, this project will become more irrelevant.

I don't think so. All recent developments of kernel back out from implementing policy protocols in the kernel. In my humble opinion _support for_ per-VLAN STP means that you can pass the STP state to the kernel, not that the full MSTP/RSTP protocol set will get implemented in-kernel.

Unless proven otherwise, I expect mstpd (or similar) project to gain much more relevance with any new related features in the kernel.

**Comment by commodo** [comment]
> > Depending on how the kernel STP evolves, this project will become more irrelevant.
> 
> I don't think so. All recent developments of kernel back out from implementing policy protocols in the kernel. In my humble opinion _support for_ per-VLAN STP means that you can pass the STP state to the kernel, not that the full MSTP/RSTP protocol set will get implemented in-kernel.
> 
> Unless proven otherwise, I expect mstpd (or similar) project to gain much more relevance with any new related features in the kernel.

Well, my opinion may be outdated.
It's from 2015, when I was thinking that maybe RSTP/MSTP may become supported in the kernel (same as STP).
That never became the case (which is fine).

It could be that mstpd [and similar] gain more traction.
Back in 2015 we were using OVS as well [because the idea was this would scale better than Linux bridges].
We also thought OVS could be used in the embedded space [boy were we wrong]; it was evolving way too quick in code-size, and you couldn't run it anymore on network devices with 128 MB of RAM [usually smart access points].
We also tried the RSTP implementation of OVS, but that was not as mature as mstpd's and it was way more difficult to integrate with switch chips.
